File: src/test/java/no/nav/tjenestepensjon/simulering/AsyncExecutorTest.java
package no.nav.tjenestepensjon.simulering;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.lessThan;

import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Executors;

import org.junit.jupiter.api.Test;

class AsyncExecutorTest {

    @Test
    void isAsyncIfMultipleThreads() {
        AsyncExecutor<String, Sleepy> executor = new AsyncExecutor<>(Executors.newFixedThreadPool(3));
        long startTime = System.currentTimeMillis();
        AsyncExecutor.AsyncResponse results = executor.executeAsync(Map.of("1", new Sleepy(), "2", new Sleepy(), "3", new Sleepy()));
        long elapsed = System.currentTimeMillis() - startTime;
        assertThat(results.getResultMap().size(), is(3));
        assertThat(elapsed, is(lessThan(3 * 100L)));
    }

    @Test
    void isSynchIfSingleThread() {
        AsyncExecutor<String, Sleepy> executor = new AsyncExecutor<>(Executors.newFixedThreadPool(1));
        long startTime = System.currentTimeMillis();
        AsyncExecutor.AsyncResponse results = executor.executeAsync(Map.of("1", new Sleepy(), "2", new Sleepy(), "3", new Sleepy()));
        long elapsed = System.currentTimeMillis() - startTime;
        assertThat(results.getResultMap().size(), is(3));
        assertThat(elapsed, is(greaterThan(3 * 100L)));
    }

    @Test
    void catchAndReturnExceptionsFromAsyncExecutions() {
        AsyncExecutor<String, Sleepy> executor = new AsyncExecutor<>(Executors.newFixedThreadPool(3));
        AsyncExecutor.AsyncResponse results = executor.executeAsync(Map.of("1", new SleepyThrowsException(), "2", new SleepyThrowsException(), "3", new Sleepy()));
        assertThat(results.getResultMap().size(), is(1));
        assertThat(results.getExceptions().size(), is(2));
    }

    private static class Sleepy implements Callable<String> {
        @Override
        public String call() throws Exception {
            Thread.sleep(100);
            return Thread.currentThread().getName();
        }
    }

    private static class SleepyThrowsException extends Sleepy implements Callable<String> {
        @Override
        public String call() {
            throw new RuntimeException("excpetion");
        }
    }
}

File: src/test/kotlin/no/nav/tjenestepensjon/simulering/AsyncExecutorTest.kt
package no.nav.tjenestepensjon.simulering

import org.junit.jupiter.api.Test
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue

internal class AsyncExecutorTest {
    @Test
    fun isAsyncIfMultipleThreads() {
        val executor: AsyncExecutor<String, Sleepy> = AsyncExecutor(Executors.newFixedThreadPool(3))
        val startTime = System.currentTimeMillis()
        val results = executor.executeAsync(mapOf("1" to Sleepy(), "2" to Sleepy(), "3" to Sleepy()))
        val elapsed = System.currentTimeMillis() - startTime
        assertEquals(3, results.resultMap.size)
        assertTrue(elapsed < 300L)
    }

    @Test
    fun isSynchIfSingleThread() {
        val executor: AsyncExecutor<String, Sleepy> = AsyncExecutor(Executors.newFixedThreadPool(1))
        val startTime = System.currentTimeMillis()
        val results = executor.executeAsync(mapOf("1" to Sleepy(), "2" to Sleepy(), "3" to Sleepy()))
        val elapsed = System.currentTimeMillis() - startTime
        assertEquals(3, results.resultMap.size)
        assertTrue(elapsed > 300L)
    }

    @Test
    fun `Catch and return exceptions from async executions`() {
        val executor: AsyncExecutor<String, Sleepy> = AsyncExecutor(Executors.newFixedThreadPool(3))
        val results = executor.executeAsync(mapOf("1" to SleepyThrowsException(), "2" to SleepyThrowsException(), "3" to Sleepy()))
        assertEquals(1, results.resultMap.size)
        assertEquals(2, results.exceptions.size)
    }

    private open class Sleepy : Callable<String> {
        @Throws(Exception::class)
        override fun call(): String {
            Thread.sleep(100)
            return Thread.currentThread().name
        }
    }

    private class SleepyThrowsException : Sleepy(), Callable<String> {
        override fun call(): String {
            throw RuntimeException("exception")
        }
    }
}